webapp/src/components/NewsList/NewsList.vue

```diff
@@ -23,49 +25,144 @@
 <script>
 import NewsItem from "../NewsItem/NewsItem.vue";
 import NewsForm from "../NewsForm/NewsForm.vue";
+import GET_POSTS from "../../graphql/Posts.gql";
+import WRITE_POST from "../../graphql/WritePost.gql";
+import DELETE_POST from "../../graphql/DeletePost.gql";
+import UPVOTE_POSTS from "../../graphql/UpvotePost.gql";
+import DOWNVOTE_POSTS from "../../graphql/DownvotePost.gql";
+
+//Use default user for now
+const DEFAULT_USER = "Jonas";
+
 export default {
+  apollo: {
+    posts: {
+      query: GET_POSTS,
+      update({ posts }) {
+        return posts;
+      },
+    },
+  },
   name: "app",
   components: {
     NewsItem,
     NewsForm,
   },
-  props: {
-    initialItems: {
-      type: Array,
-      default: () => {
-        return [
-          { id: 1, title: "VueJS", votes: 0 },
-          { id: 2, title: "Hello world!", votes: 0 },
-        ];
-      },
-    },
-  },
   data: function () {
     return {
-      items: [...this.initialItems],
+      posts: [],
       descending: true,
     };
   },
   methods: {
-    updateItem: function (item) {
-      if (item && item.id) {
-        this.items = this.items.filter((el) => el.id !== item.id);
-        this.items = [...this.items, item];
-      }
+    upvote: function (post) {
+      this.$apollo.mutate({
+        mutation: UPVOTE_POSTS,
+        variables: {
+          id: post.id,
+          voter: { name: DEFAULT_USER },
+        },
+        update: (store, { data: { upvote } }) => {
+          const data = store.readQuery({ query: GET_POSTS });
+          let upvotedElement = data.posts.find((el) => el.id === upvote.id);
+          if (upvotedElement) upvotedElement = upvote;
+          store.writeQuery({ query: GET_POSTS, data });
+        },
+        optimisticResponse: {
+          upvote: {
+            __typename: "Post",
+            id: post.id,
+            title: post.title,
+            votes: post.votes + 1,
+            author: {
+              __typename: "User",
+              name: DEFAULT_USER,
+            },
+          },
+        },
+      });
     },
-    removeItem: function (item) {
-      if (item && item.id) {
-        this.items = this.items.filter((el) => el.id !== item.id);
-      }
+    downvote: function (post) {
+      this.$apollo.mutate({
+        mutation: DOWNVOTE_POSTS,
+        variables: {
+          id: post.id,
+          voter: { name: DEFAULT_USER },
+        },
+        update: (store, { data: { downvote } }) => {
+          const data = store.readQuery({ query: GET_POSTS });
+          let downvotedElement = data.posts.find((el) => el.id === downvote.id);
```

Comment:
Are you sure that this is even necessary? As far as I know, `vue-apollo` is smart enough to update existing types in the cache via the `ID`! That's why the `ID` should always be requested.

Reply:
Try to omit this and give me feedback if it has the desired effect or not.

Reply:
You are right! It is indeed not necessary here. Only for write and delete it needs to be updated manually.

Reply:
`vue-apollo` is really smart, hugh? :wink: That's actually one of the most amazing benefits that comes with a typed schema, the client can cleverly auto-update it's cache and you don't have to do anything yourself.

Reply:
You can also install apollo client browser plugin (available for Firefox and Chromium) and observe the cache in real time.

Reply:
That sounds awesome! Will definitely check it out! 😄